collector/node_collector.go

```diff
@@ -52,58 +60,59 @@ type metricsMeta struct {
 // NewAgent returns a new instance of a DC/OS agent poller based on the provided
 // configuration and the result of the provided ipCommand script for detecting
 // the agent's IP address.
-func NewAgent(ipCommand string, port int, pollPeriod time.Duration, metricsChan chan<- producers.MetricsMessage) (Agent, error) {
-	a := Agent{}
+func NewDCOSHost(dcosRole string, ipCommand string, port int, pollPeriod time.Duration, metricsChan chan<- producers.MetricsMessage) (DCOSHost, error) {
+	h := DCOSHost{}
 
 	if len(ipCommand) == 0 {
-		return a, fmt.Errorf("Must pass ipCommand to NewAgent()")
+		return h, fmt.Errorf("Must pass ipCommand to DCOSHost()")
 	}
 	if port < 1024 {
-		return a, fmt.Errorf("Must pass port >= 1024 to NewAgent()")
+		return h, fmt.Errorf("Must pass port >= 1024 to DCOSHost()")
 	}
 	if pollPeriod == 0 {
-		return a, fmt.Errorf("Must pass pollPeriod to NewAgent()")
+		return h, fmt.Errorf("Must pass pollPeriod to DCOSHost()")
 	}
 
-	a.IPCommand = ipCommand
-	a.Port = port
-	a.PollPeriod = pollPeriod
-	a.MetricsChan = metricsChan
+	h.IPCommand = ipCommand
+	h.Port = port
+	h.PollPeriod = pollPeriod
+	h.MetricsChan = metricsChan
+	h.DCOSRole = dcosRole
 
 	// Detect the agent's IP address once. Per the DC/OS docs (at least as of
 	// November 2016), changing a node's IP address is not supported.
 	var err error
-	if a.AgentIP, err = a.getIP(); err != nil {
-		return a, err
+	if h.IPAddress, err = h.getIP(); err != nil {
+		return h, err
 	}
 
-	return a, nil
+	return h, nil
 }
 
 // RunPoller periodiclly polls the HTTP APIs of a Mesos agent. This function
 // should be run in its own goroutine.
-func (a *Agent) RunPoller() {
-	ticker := time.NewTicker(a.PollPeriod)
+func (h *DCOSHost) RunPoller() {
+	ticker := time.NewTicker(h.PollPeriod)
 
 	// Poll once immediately
-	for _, m := range a.transform(a.pollAgent()) {
-		a.MetricsChan <- m
+	for _, m := range h.transform(h.pollHost()) {
+		h.MetricsChan <- m
 	}
 	for {
 		select {
 		case _ = <-ticker.C:
-			for _, m := range a.transform(a.pollAgent()) {
-				a.MetricsChan <- m
+			for _, m := range h.transform(h.pollHost()) {
+				h.MetricsChan <- m
 			}
 		}
 	}
 }
 
 // getIP runs the ip_detect script and returns the IP address that the agent
 // is listening on.
-func (a *Agent) getIP() (string, error) {
-	log.Debugf("Executing ip-detect script %s", a.IPCommand)
-	cmdWithArgs := strings.Split(a.IPCommand, " ")
+func (h *DCOSHost) getIP() (string, error) {
+	nodeColLog.Debugf("Executing ip-detect script %s", h.IPCommand)
+	cmdWithArgs := strings.Split(h.IPCommand, " ")
 
 	ipBytes, err := exec.Command(cmdWithArgs[0], cmdWithArgs[1:]...).Output()
```

Comment:
please don't call `exe.Command(...).Output()` this is dangerous. I'd rather use https://github.com/dcos/dcos-go/tree/master/exec or wait for https://github.com/dcos/dcos-go/pull/15 it should be landing today. It will give you a safer way of executing detect_ip script with result validation